lang/core/iter/range.go

```diff
@@ -0,0 +1,262 @@
+	if typ.Kind != types.KindFunc {
+		return nil, fmt.Errorf("must be of kind func")
+	}
+
+	if len(typ.Ord) < 1 || len(typ.Ord) > 3 {
+		return nil, fmt.Errorf("the range function needs one to three args")
+	}
+
+	if typ.Out == nil {
+		return nil, fmt.Errorf("return type of function must be specified")
+	}
+
+	if typ.Out.Cmp(types.NewType("[]int")) != nil {
+		return nil, fmt.Errorf("return type of function must be a list of ints")
+	}
+
+	obj.Type = typ.Copy() // this is to store the type of return value
+
+	return obj.Type, nil
+}
+
+// Info returns some static info about itself. Build must be called before this
+// will return correct data
+func (obj *RangeFunc) Info() *interfaces.Info {
+	return &interfaces.Info{
+		Pure: true,
+		Memo: false,
+		Sig:  obj.Type,
+		Err:  obj.Validate(),
+	}
+}
+
+// Validate tells us if the input struct takes a valid form.
+func (obj *RangeFunc) Validate() error {
+	if obj.Type == nil {
+		return fmt.Errorf("must specify a type")
+	}
+	return nil
+}
+
+// Init runs some startup code for this function.
+func (obj *RangeFunc) Init(init *interfaces.Init) error {
+	obj.init = init
+	return nil
+}
+
+// Stream returns the changing values that this func has over time.
+func (obj *RangeFunc) Stream(ctx context.Context) error {
+	defer close(obj.init.Output) // closing the sender
+
+	for {
+		select {
+		case input, ok := <-obj.init.Input:
+			if !ok {
+				return nil // we don't have more inputs
+			}
+
+			if obj.last != nil && input.Cmp(obj.last) == nil {
+				continue // nothing has changed, skip it
+			}
+			obj.last = input // storing the input for comparison
+
+			args, err := interfaces.StructToCallableArgs(input)
+			if err != nil {
+				return err
+			}
+
+			result, err := obj.Call(ctx, args)
+			if err != nil {
+				return err
+			}
+
+			if obj.result != nil && result.Cmp(obj.result) == nil {
+				continue // if the result didn't change, we don't need to update
+			}
+			obj.result = result // store new result
+
+		case <-ctx.Done():
+			return nil
+		}
+
+		select {
+		case obj.init.Output <- obj.result: // sending new result
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// Call returns the result of this function.
+func (obj *RangeFunc) Call(ctx context.Context, args []types.Value) (types.Value, error) {
+	var result []types.Value
+	var err error
+	if len(args) == 1 { // we only have stop, assume start is 0 and step is 1
+		result, err = obj.run(ctx, 0, args[0].Int(), 1)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if len(args) == 2 { // we have start and stop, assume step is 1
+		result, err = obj.run(ctx, args[0].Int(), args[1].Int(), 1)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if len(args) == 3 { // we have all the args
+		result, err = obj.run(ctx, args[0].Int(), args[1].Int(), args[2].Int())
+		if err != nil {
+			return nil, err
+		}
+	}
+	// TODO: should call the for function here and return the result
+	return &types.ListValue{V: result}, nil
```

Comment:
Why not build this in run() and return directly above?
Also look at some examples, it wasn't built correctly, it needs a type.